src/librustc/ty/mod.rs

```diff
@@ -1726,6 +1728,190 @@ impl<'tcx, 'container> AdtDefData<'tcx, 'container> {
+                let adt_ty =
+                    adt.sized_constraint
+                    .unwrap(DepNode::SizedConstraint(adt.did))
+                    .subst(tcx, substs);
+                debug!("sized_constraint_for_ty({:?}) intermediate = {:?}",
+                       ty, adt_ty);
+                self.sized_constraint_for_ty(tcx, stack, adt_ty)
+            }
+
+            TyProjection(..) => {
+                // must calculate explicitly.
+                // FIXME: consider special-casing always-Sized projections
+                ty
+            }
+
+            TyParam(..) => {
+                // perf hack: if there is a `T: Sized` bound, then
+                // we know that `T` is Sized and do not need to check
+                // it on the impl.
+
+                let sized_trait = match tcx.lang_items.sized_trait() {
+                    Some(x) => x,
+                    _ => return ty
+                };
+                let sized_predicate = Binder(TraitRef {
+                    def_id: sized_trait,
+                    substs: tcx.mk_substs(Substs::new_trait(
+                        vec![], vec![], ty
+                    ))
+                }).to_predicate();
+                let predicates = tcx.lookup_predicates(self.did).predicates;
+                if predicates.into_iter().any(|p| p == sized_predicate) {
+                    tcx.types.bool
+                } else {
+                    ty
+                }
+            }
+
+            TyInfer(..) => {
+                bug!("unexpected type `{:?}` in sized_constraint_for_ty",
+                     ty)
+            }
+        };
+        debug!("sized_constraint_for_ty({:?}) = {:?}", ty, result);
+        result
+    }
+
+    /// Calculates the Sized-constraint.
+    ///
+    /// As the Sized-constraint of enums can be a *set* of types,
+    /// the Sized-constraint may need to be a set also. Because introducing
+    /// a new type of IVar is currently a complex affair, the Sized-constraint
+    /// may be a tuple.
+    ///
+    /// In fact, there are only a few options for the constraint:
+    ///     - `bool`, if the type is always Sized
+    ///     - an obviously-unsized type
+    ///     - a type parameter or projection whose Sizedness can't be known
+    ///     - a tuple of type parameters or projections, if there are multiple
+    ///       such.
+    ///     - a TyError, if a type contained itself. The representability
+    ///       check should catch this case.
+    fn calculate_sized_constraint_inner(&'tcx self, tcx: &ty::TyCtxt<'tcx>,
+                                        stack: &mut Vec<AdtDefMaster<'tcx>>)
+    {
+
+        let dep_node = DepNode::SizedConstraint(self.did);
+
+        if self.sized_constraint.get(dep_node).is_some() {
+            return;
+        }
+
+        if stack.contains(&self) {
+            debug!("calculate_sized_constraint: {:?} is recursive", self);
+            self.sized_constraint.fulfill(dep_node, tcx.types.err);
```

Comment:
I feel mildly nervous to return `err` here when, in fact, no error has been reported. This has gotten us into some trouble before. I think this case is different, though, in that this is kind of a special function, and this type maybe doesn't escape into "general use" -- certainly it's somewhat reasonable to impose on the callers that they should report an error or handle `err` carefully. 

Another possible concern is that we may get false error reports -- perhaps the final type in the struct is actually `ty::err`? In that case, the return type would indicate recursion, potentially generating a false error report -- somewhat ironically, since `ty::err` is supposed to *suppress* error reports.

I guess what I'm getting as: maybe better to return `Result<Ty, Recursive>` (where `struct Recursive;`) instead of just `Ty` (or else `Option<Ty>`, as you prefer).